### PR TITLE
[Merged by Bors] - feat: pushouts of monoids

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/PushoutI.lean [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+in the diagram are injective).
+
+## Main definitions
+
+- `Monoid.PushoutI`: the pushout of a diagram of monoids indexed by a type `ι`
+- `Monoid.PushoutI.base`: the map from the amalgamating monoid to the pushout
+- `Monoid.PushoutI.of`: the map from each Monoid in the family to the pushout
+- `Monoid.PushoutI.lift`: the universal property used to define homomorphisms out of the pushout.
+
+- `Monoid.PushoutI.NormalWord`: a normal form for words in the pushout
```
Comment (riccardobrasca) [inline Mathlib/GroupTheory/PushoutI.lean:27]: This is in the second part.

**Review by riccardobrasca — APPROVED** [review]
Can you add the universal property in some form? It is essentially already there with `lift` and `hom_ext_nonempty`, but it would be nice to have bijection from maps from the pushout and maps such that...

**Comment by ChrisHughes24** [comment]
> Can you add the universal property in some form? It is essentially already there with `lift` and `hom_ext_nonempty`, but it would be nice to have bijection from maps from the pushout and maps such that...

I added it, but I never end up using these things. The full universal property is given by `of`, `base`, of_comp_eq_base`, `lift` and `hom_ext`.

**Comment by riccardobrasca** [comment]
> > Can you add the universal property in some form? It is essentially already there with `lift` and `hom_ext_nonempty`, but it would be nice to have bijection from maps from the pushout and maps such that...
> 
> I added it, but I never end up using these things. The full universal property is given by `of`, `base`, of_comp_eq_base`, `lift`and`hom_ext`.

I agree that in practice one uses the explicit results, but I think it's nice to have it, since it really packages the universal property. Can you just add a TODO that says we want to add `HasPushout` (or whatever it is called) to the relevant categories? Thanks!

bors d+

**Comment by ChrisHughes24** [comment]
> > > Can you add the universal property in some form? It is essentially already there with `lift` and `hom_ext_nonempty`, but it would be nice to have bijection from maps from the pushout and maps such that...
> > 
> > 
> > I added it, but I never end up using these things. The full universal property is given by `of`, `base`, of_comp_eq_base`, `lift`and`hom_ext`.
> 
> I agree that in practice one uses the explicit results, but I think it's nice to have it, since it really packages the universal property. Can you just add a TODO that says we want to add `HasPushout` (or whatever it is called) to the relevant categories? Thanks!
> 
> bors d+

We already know that both Monoids and Groups have all colimits
bors r+

**Pull request closed without merging**